processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -0,0 +1,195 @@
+			}
+		}
+	}
+
+	return pdatautil.MetricsFromMetricsData(mds)
+}
+
+// update updates the original metric content in the metricPtr pointer.
+func (mtp *metricsTransformProcessor) update(metricPtr *metricspb.Metric, transform Transform) {
+	// metric name update
+	if transform.NewName != "" {
+		metricPtr.MetricDescriptor.Name = transform.NewName
+	}
+
+	for _, op := range transform.Operations {
+		// update label
+		if op.Action == UpdateLabel {
+			mtp.updateLabelOp(metricPtr, op)
+		}
+		// aggregate across labels
+
+		// aggregate across label values
+
+	}
+}
+
+// insert inserts a new copy of the metricPtr content into the metricPtrs slice.
+// returns the new metrics list and the new metric
+func (mtp *metricsTransformProcessor) insert(metricPtr *metricspb.Metric, metricPtrs []*metricspb.Metric, transform Transform) ([]*metricspb.Metric, *metricspb.Metric) {
+	metricCopy := mtp.createCopy(metricPtr)
+	mtp.update(metricCopy, transform)
+	return append(metricPtrs, metricCopy), metricCopy
+}
+
+func (mtp *metricsTransformProcessor) updateLabelOp(metricPtr *metricspb.Metric, op Operation) {
+	// if new_label is invalid, skip this operation
+	if !mtp.validNewLabel(metricPtr.MetricDescriptor.LabelKeys, op.NewLabel) {
+		log.Printf("error running %q processor due to collided %q: %v with existing label on metric named: %v detected by function %q", typeStr, NewLabelFieldName, op.NewLabel, metricPtr.MetricDescriptor.Name, validNewLabelFuncName)
+		return
+	}
+
+	for _, label := range metricPtr.MetricDescriptor.LabelKeys {
+		if label.Key != op.Label {
+			continue
+		}
+		// label key update
+		if op.NewLabel != "" {
+			label.Key = op.NewLabel
+		}
+		// label value update
+	}
+}
+
+// createCopy creates a new copy of the input metric.
+func (mtp *metricsTransformProcessor) createCopy(metricPtr *metricspb.Metric) *metricspb.Metric {
+	copyMetricDescriptor := *metricPtr.MetricDescriptor
+	copyLabelKeys := make([]*metricspb.LabelKey, 0)
+	for _, labelKey := range copyMetricDescriptor.LabelKeys {
+		copyLabelKeys = append(
+			copyLabelKeys,
+			&metricspb.LabelKey{
+				Key:         labelKey.Key,
+				Description: labelKey.Description,
+			},
+		)
+	}
+	copyMetricDescriptor.LabelKeys = copyLabelKeys
+
+	copy := &metricspb.Metric{
+		MetricDescriptor: &copyMetricDescriptor,
+		Timeseries:       metricPtr.Timeseries,
+		Resource:         metricPtr.Resource,
+	}
+	return copy
+}
+
+// validNewName determines if the new name is a valid one. An invalid one is one that already exists.
+func (mtp *metricsTransformProcessor) validNewName(transform Transform, nameToMetricMapping map[string]*metricspb.Metric) bool {
+	_, ok := nameToMetricMapping[transform.NewName]
+	return !ok
+}
+
+// validNewLabel determines if the new label is a valid one. An invalid one is one that already exists.
+func (mtp *metricsTransformProcessor) validNewLabel(labelKeys []*metricspb.LabelKey, newLabel string) bool {
```

Comment:
This code is relatively trivial and straightforward, but even still, similar to above, I think I'm a fan of doing minimal validation here and allowing this processor to be lean and fast. If the user screws up, they will export bad data, or more likely get an error at export time.

Happy to be convinced otherwise on this one though.

Reply:
This would be for the same reasoning as the other validations. (This one is to guard against renaming a label to a name that already exists because this should be done through an aggregation) These are all to keep the boundaries of the expected behaviors, and once the operations step out of bounds, the processor is able to notice. However, after thinking about the single responsibility of the processor, I do see how these validations should be removed. Since the processors should only worry about transforming the data specified by the user even the user's specification might imply a possible error. I also don't see any other processors that validate data processing,  so from a convention stand point, I also think removing the validations might be reasonable. However, this also means that users may have data transformed in ways that might result in bad output metric data. If this is ok for processors, than with your approval, I would love to remove the validations because this also makes the code a lot more straightforward.

Reply:
My opinion is lets start with not having it and keep the code very simple. We can add it in later if people want it. But I'm okay with leaving it there if you feel strongly about it

Reply:
No problem! The current plan is removing the validations, but I will keep this code saved in another branch. If it is needed later, we will put this back in.